homeassistant/components/tuya/__init__.py

```diff
@@ -56,44 +65,50 @@
 )
 
 
-def setup(hass, config, retry_delay=FIRST_RETRY_TIME):
-    """Set up Tuya Component."""
-
-    _LOGGER.debug("Setting up integration")
+async def async_setup(hass, config):
+    """Set up the Tuya integration."""
 
-    tuya = TuyaApi()
-    username = config[DOMAIN][CONF_USERNAME]
-    password = config[DOMAIN][CONF_PASSWORD]
-    country_code = config[DOMAIN][CONF_COUNTRYCODE]
-    platform = config[DOMAIN][CONF_PLATFORM]
+    conf = config.get(DOMAIN)
+    if conf is not None:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+            )
+        )
 
-    try:
-        tuya.init(username, password, country_code, platform)
-    except (TuyaNetException, TuyaServerException):
+    return True
 
-        _LOGGER.warning(
-            "Connection error during integration setup. Will retry in %s seconds",
-            retry_delay,
-        )
 
-        def retry_setup(now):
-            """Retry setup if a error happens on tuya API."""
-            setup(hass, config, retry_delay=min(2 * retry_delay, MAX_RETRY_TIME))
+async def async_setup_entry(hass, entry):
+    """Set up Tuya platform."""
 
-        call_later(hass, retry_delay, retry_setup)
+    tuya = TuyaApi()
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    country_code = entry.data[CONF_COUNTRYCODE]
+    platform = entry.data[CONF_PLATFORM]
 
-        return True
+    try:
+        await hass.async_add_executor_job(
+            tuya.init, username, password, country_code, platform
+        )
+    except (TuyaNetException, TuyaServerException) as exc:
+        raise ConfigEntryNotReady() from exc
 
     except TuyaAPIException as exc:
         _LOGGER.error(
             "Connection error during integration setup. Error: %s", exc,
         )
         return False
 
-    hass.data[DATA_TUYA] = tuya
-    hass.data[DOMAIN] = {"entities": {}}
+    hass.data[TUYA_DATA] = tuya
```

Comment:
Ideally all of these are stored under `hass.data[DOMAIN][entry.entry_id]` 

Something like (renaming probably needed)... 
```
hass.data[DOMAIN][entry.entry_id] = {
    CONFIG_ENTRY_IS_SETUP: set(),
    TUYA_DATA: tuya,
    "entities": {},
    "pending": {},
}
```

Reply:
For the moment I moved all under [DOMAIN] level, entry_id require more change and for the moment is not necessary (single instance integration)